### PR TITLE
Deprecate ViewHandler getters in favor of constructor injections

### PR DESCRIPTION
**Pull request by Ener-Getick**
See https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1287#issuecomment-167755837

### CODE REVIEW

**Comment by Ener-Getick** [comment]
Should I use the same method used [here](https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1252/files#diff-afaa78b4ed090b310e8499c8ecd788f5R112) to trigger a deprecation?

**Comment by xabbuh** [comment]
Isn't it sufficient to simply always trigger the deprecation when one of these methods is called?

**Comment by Ener-Getick** [comment]
The bundle would trigger deprecations that can't be fixed so I don't really like that...

**Comment by xabbuh** [comment]
In Symfony we had a few places where we passed a special argument to silence deprecations in a method. But what are the parts where we still call these methods? And is that really necessary?

**Comment by Ener-Getick** [comment]
We always call this methods in the ``ViewHandler`` when we need to access the serializer, the router or the template engine.
I'll use the reflection because in the case these methods are extended nothing will be triggered with your method (it seems that they are only there to be extended not to be called).

**Comment by Ener-Getick** [comment]
Done. 

@lsmith77 @xabbuh would be amazing to release ``1.8`` and ``2.0`` before the end of the year! We have almost finished.

**Comment by xabbuh** [comment]
:+1: LGTM

File: View/ViewHandler.php [hunk 1]
```diff
@@ -533,4 +546,19 @@ private function getDataFromView(View $view)
...
+    /**
+     * Triggers a deprecation if a getter is extended.
+     *
+     * @todo remove this in 2.0.
+     */
+    private function deprecateGetter($name)
+    {
+        if (is_subclass_of($this, __CLASS__)) {
+            $method = new \ReflectionMethod($this, $name);
+            if ($method->getDeclaringClass()->getName() != __CLASS__) {
```
Comment (stof) [inline View/ViewHandler.php]: this should use strict comparison

File: View/ViewHandler.php [hunk 2]
```diff
@@ -533,4 +546,19 @@ private function getDataFromView(View $view)
...
+                @trigger_error(sprintf('%s::%s() is deprecated since version 1.8 and will be removed in 2.0.', __CLASS__, $name), E_USER_DEPRECATED);
```
Comment (stof) [inline View/ViewHandler.php]: You should mention that overwriting it is deprecated (and maybe mention ``get_class($this)`` in the message too to help finding where it happens)

Comment (Ener-Getick) [inline View/ViewHandler.php]: Is it better?

File: View/ViewHandler.php (hunk 1 above)
Comment (Ener-Getick) [inline View/ViewHandler.php]: Oops good catch

**Comment by lsmith77** [comment]
I guess we didn't manage that and I also think we should do an RC phase before we make the final releases. As for this PR .. fine for me .. though I do not have so much experience with managing deprecations

**Comment by xabbuh** [comment]
@stof Do you see anything else we may want to change? Otherwise this look good to merge to me.

**Pull request merged**